lib/user_name_suggester.rb

```diff
@@ -39,28 +41,51 @@ def self.fix_username(name)
   end
 
   def self.sanitize_username(name)
-    name = ActiveSupport::Inflector.transliterate(name.to_s)
-    # 1. replace characters that aren't allowed with '_'
+    name = name.to_s
+
+    if SiteSetting.unicode_usernames
+      name.unicode_normalize!
+    else
+      name = ActiveSupport::Inflector.transliterate(name)
+    end
+
+    name.gsub!(UsernameValidator.invalid_char_pattern, '_')
+    name.chars.map! { |c| UsernameValidator.whitelisted_char?(c) ? c : '_' } if UsernameValidator.char_whitelist_exists?
+    name.gsub!(UsernameValidator::INVALID_LEADING_CHAR_PATTERN, '')
     name.gsub!(UsernameValidator::CONFUSING_EXTENSIONS, "_")
-    name.gsub!(/[^\w.-]/, "_")
-    # 2. removes unallowed leading characters
-    name.gsub!(/^\W+/, "")
-    # 3. removes unallowed trailing characters
-    name = remove_unallowed_trailing_characters(name)
-    # 4. unify special characters
-    name.gsub!(/[-_.]{2,}/, "_")
+    name.gsub!(UsernameValidator::INVALID_TRAILING_CHAR_PATTERN, '')
+    name.gsub!(UsernameValidator::REPEATED_SPECIAL_CHAR_PATTERN, '_')
     name
   end
 
-  def self.remove_unallowed_trailing_characters(name)
-    name.gsub!(/[^A-Za-z0-9]+$/, "")
+  def self.rightsize_username(name)
+    name = truncate(name, User.username_length.end)
+    name.gsub!(UsernameValidator::INVALID_TRAILING_CHAR_PATTERN, '')
+
+    missing_char_count = User.username_length.begin - name.grapheme_clusters.size
+    name << '1' * missing_char_count if missing_char_count > 0
```

Comment:
`1` is an interesting padding character. Probably fine.

Reply:
I think "1" has been used as a padding character for a very long time. I can replace it with something fancier in a follow-up commit if you have a good suggestion.